### PR TITLE
[MPS] Add support for complex scalars

### LINKED ISSUE #119088: complex tensors do not work on MPS
### 🐛 Describe the bug

On a Mac (macOS Monterrey Apple M1 Pro, full system details below):

```python
import torch

rand_tensor = torch.rand((2,2)).to('mps')
rand_tensor * 1j
```

This produces:

```
RuntimeError: c10::isComplexType(input.scalar_type()) && c10::isComplexType(other.scalar_type()) INTERNAL ASSERT FAILED at "/Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/native/mps/operations/BinaryKernel.mm":340, please report a bug to PyTorch.
```

### Versions

```
PyTorch version: 2.2.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.13 (main, Sep 11 2023, 08:16:02) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-12.5.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] pytorch-lightning==2.1.4
[pip3] torch==2.2.0
[pip3] torchaudio==2.2.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.3.0.post0
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] pytorch-lightning         2.1.4                    pypi_0    pypi
[conda] torch                     2.2.0                    pypi_0    pypi
[conda] torchaudio                2.2.0                    pypi_0    pypi
[conda] torchinfo                 1.8.0                    pypi_0    pypi
[conda] torchmetrics              1.3.0.post0              pypi_0    pypi
```

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr

### PR DESCRIPTION
**Pull request by malfet**
- Switch to native complex support if running on MacOS Monterey or newer for binary ops.
- Python complex scalars are always represented in PyTorch as ComplexDouble, but MPS yet to support double precision types, so downcast them to floats
- Also add `cf`(for complex float)  and `ch`(for complex half) to MPSScalar value union
- Fix complex scalars to view promotion, by introducing `legacy_complex_as_view` helper function, that non-float types to complex and promotes CPU complex scalars to MPS before turning them into a view.
- Add `test_tensor_scalar_binops`

Fixes https://github.com/pytorch/pytorch/issues/119088

Test plan: CI (have quite a lot of tests, see new unexpected successes) +  `python -c "import torch;x,y=torch.rand(2, 2, dtype=torch.cfloat, device='mps'),torch.tensor(2+3j,dtype=torch.chalf);print(y+x)"`

### CODE REVIEW

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 1]
```diff
@@ -414,6 +417,11 @@ MPSScalar getMPSScalar(const Scalar& scalar, ScalarType type) {
       return {.value.i = scalar.to<uint8_t>(), .size = sizeof(uint8_t), .type = type};
     case ScalarType::Bool:
       return {.value.b = scalar.to<bool>(), .size = sizeof(bool), .type = type};
+    case ScalarType::ComplexHalf:
+      return {.value.cf = scalar.to<c10::complex<at::Half>>(), .size = sizeof(int32_t), .type = type};
```
Comment (albanD) [inline aten/src/ATen/native/mps/OperationUtils.mm:421]: .value.ch?

**Review by albanD — APPROVED** [review]
Nice!

**Comment by albanD** [comment]
More unexpected success I guess ;)

**Comment by malfet** [comment]
> More unexpected success I guess ;)
No, it's fp32 + complex64 binary ops failures (which was not tested previously and what the issue is about)

**Pull request closed without merging**